eng/scripts/Pack-Crates.ps1

```diff
@@ -126,26 +126,65 @@ function Get-PackagesToBuild() {
   return $buildOrder
 }
 
-function Initialize-VendorDirectory() {
-  $path = "$RepoRoot/target/vendor"
-  Invoke-LoggedCommand "cargo vendor $path" -GroupOutput | Out-Host
-  return $path
-}
+# https://doc.rust-lang.org/cargo/reference/registry-web-api.html#publish
+# https://github.com/rust-lang/cargo/blob/5c87c14f9a162daf10d4133fdaab35c72d67b018/crates/crates-io/lib.rs#L42
+function Get-ApiMetadata($package) {
+  $packagePath = Split-Path -Path $package.manifest_path -Parent
+  $readmePath = $package.readme ? (Join-Path -Path $packagePath -ChildPath $package.readme) : $null
+
+  $jsonBody = [ordered]@{
+    'name'          = $package.name
+    'vers'          = $package.version
+    'deps'          = @()
+    'features'      = $package.features
+    'authors'       = $package.authors
+    'description'   = $package.description
+    'documentation' = $package.documentation
+    'homepage'      = $package.homepage
+    'readme'        = if ($readmePath -and (Test-Path -Path $readmePath)) {
+      Get-Content -Path $readmePath -Raw
+    }
+    else {
+      $null
+    }
+    'readme_file'   = $package.readme
+    'keywords'      = $package.keywords
+    'categories'    = $package.categories
+    'license'       = $package.license
+    'license_file'  = $package.license_file
+    'repository'    = $package.repository
+    'links'         = $package.links
+    'rust_version'  = $package.rust_version
+  }
 
-function Add-CrateToLocalRegistry($LocalRegistryPath, $Package) {
-  $packageName = $Package.name
-  $packageVersion = $Package.version
+  foreach ($dependency in $package.dependencies) {
+    $jsonBody.deps += @{
+      'name'                  = $dependency.name
+      'version_req'           = $dependency.req
+      'features'              = $dependency.features
+      'optional'              = $dependency.optional
+      'default_features'      = $dependency.default_features
+      'target'                = $dependency.target
+      'kind'                  = $dependency.kind
+      'explicit_name_in_toml' = $dependency.rename
+    }
+  }
 
-  # create an index entry for the package
-  $packagePath = "$RepoRoot/target/package/$packageName-$packageVersion"
+  return $jsonBody
+}
 
-  Write-Host "Copying package '$packageName' to vendor directory '$LocalRegistryPath'"
-  Copy-Item -Path $packagePath -Destination $LocalRegistryPath -Recurse
+function New-ApiPutFile($crateMetadata, $crateFilePath) {
+  $metadataBytes = [Text.Encoding]::Utf8.GetBytes($crateMetadata)
+  $metadataLengthBytes = [BitConverter]::GetBytes([UInt32]$metadataBytes.Length)
+  $crateBytes = [IO.File]::ReadAllBytes($crateFilePath)
+  $crateLengthBytes = [BitConverter]::GetBytes([UInt32]$crateBytes.Length)
 
-  #write an empty checksum file
-  '{"files":{}}' | Out-File -FilePath "$LocalRegistryPath/$packageName-$packageVersion/.cargo-checksum.json" -Encoding utf8
+  $bytes += $metadataLengthBytes + $metadataBytes + $crateLengthBytes + $crateBytes
+
+  return $bytes
 }
```

Comment:
Why not just use `Get-FileHash`? It's been around forever. Best not to reinvent the wheel and own its maintenance.

Reply:
this isn't computing a hash.  Its constructing a binary file for posting to the crates.io api

Reply:
...which comes back to my original question: why are we doing that? I get why we crack open `.nupkg` files: there's binaries to check for signing, viruses, etc. `.crate` files are zipped source. We spending so much time on redoing what `cargo publish` already does which not only prevents some other work from getting done, but I worry creates a dependency on undocumented implementation details we'll forever have to maintain ourselves.